### ISSUE TITLE
Multipart handling

### ISSUE THREAD

**Issue by jugglerchris** [lumail/lumail#248]
There have been previous issues (eg #114) about multipart MIME handling.  I still think it's not quite right.  :-)

Specifically, I think that instead of a flat lists of attachments, there should be a hierarchical tree, and that multipart/mixed need to be treated differently to multipart/alternative by default.

I occasionally receive some messages which seem empty; it turns out that they look like:
* multipart/mixed
  * text/html (this is the main body of the message)
  * text/plain (a .signature, added by a mailing list in this case)

lumail is only showing the signature, so the message appears empty. If it were multipart/alternative then I think it would be doing the right thing - but with multipart/mixed all the parts are important.

If the full attachment tree is available to Lua, then at least the body callbacks can make intelligent decisions, even if the core doesn't want to implement the logic.

**Comment by skx** [comment]
Yes I think there are faults here.  I've kinda stalled though, since it seems there is a big decision to be made:

* Continue making incremental updates and leave thing as-is.
* Restart the project, using the experience I've gained, and making **everything** an object.

The second approach is really appealing, and I made a toy-prototype at [skx/lumail2](https://github.com/skx/lumail2), but the obvious downside is that it's a fair amount of work, would break most of the documentation and examples, and would take a lot of time.

Mitigation though?  A lot of the recent pull-requests have been trying to subvert the current approach and slip in objects by the back-door.  Really suggesting that reworking and considering Maildirs, messages, and collections as first-class objects is really what people want, and a more obvious way to design things.

In short - the client works as-is for me, but I appreciate that people get different kinds of mails, and I think that there needs to be an overhaul.  I just don't know if it should be:

* Complete.  Strip things down, make everything an object, reconsidering the API.
* Minor.  Just fix bugs as and when they're reported.  Incrementally improving the existing MIME support on a per-bug-report issue.

Feedback welcome, but .. until I can decide I think I'm unlikely to make any commits in the near future.

**Comment by jugglerchris** [comment]
My view is:
* The idea of lumail2 looks/sounds exactly what I think a mail program should be.
* ...but the current lumail is good enough for me to use as my primary mail program *now*
* ...and there are plenty of incremental ways of making it even better, and I seem to tackle one every few months or so.

The option that works best for me is to evolve the current lumail towards the lumail2 ideal.  I don't mind incompatible changes along the way to move towards an improved design, but I probably wouldn't be able to put much effort into a lumail2 until it's at a similar level of functionality as lumail (err, plus some stuff I haven't fed back yet that I'm using locally) is now.

If there were a written design for lumail2 (eg proposed object interfaces for maildirs/messages/attachments/etc., but maybe also UI stuff), then I would probably (no guarantees of course :-) ) end up doing some it inside lumail1.

For example in the short term I'm interested in getting access to the full MIME tree from Lua; if there were a design for how you get that from the lumail2 message objects I would aim to add that to the lumail1 message class (at least as exposed to Lua); the old CAttachment list method could stay there until the rest of the code transitioned to the new method.

(End rambling for now)

**Comment by skx** [comment]
Your three-point summary is pretty much how I feel too:

* lumail2 is perfect.
* lumail is good enough as-is.
* lumail can probably become lumail2 easily.

The attachment/mime changes seem so fundamental to everything else that I'd lean towards prototyping them in the lumail2 repository.  Right now I have no API-design in mind, but it seems obvious that a message should contain a table of headers, then a collection of MessageParts, which will correspond to the MIME parts. Some/all of those might be attachments.

Really it comes down to guessing what people want to do with the parts/attachments as much as anything else.

Tonight I'll update lumail2 with some new code to handle attachments, and parts, and see how it goes.  If it seems to be easy/obvious/natural it can be built on.  If it is too hard, too fiddly, or too broken then I'll abandon the project and stick with lumail.

**Comment by skx** [comment]
Just thinking out loud, but I think this is what I want to implement:

* CMessage::get_parts
    * Will return an array of CMessageParts
* A CMessagePart object just contains
    * is_attachment() -> true/false
    * content_type() -> text/plain, etc.
    * content() -> The content of the part.

From the lua-side we'll have message::parts() which will return an array.  Each array element will be a CMessagePart wrapper, providing accessors to the corresponding fields.

Does that seem sane to you?

**Comment by jugglerchris** [comment]
The aspect I don't think that captures is nested MIME parts.  A multipart/mixed can include a text/plain and a multipart/alternative, which itself contains a text/plain and a text/html.  I think that structure all needs to be available somehow.  Also, you might have an attachment which is a forwarded-as-MIME e-mail with its own parts.

Unless you mean `CMessage::get_parts` only returns the top-level parts (which might be a single multpart/foo), and the multipart/foo parts also have a `get_parts` (or `get_subparts`, or whever) method.

**Comment by skx** [comment]
I was only thinking of top-level parts.  Certainly I don't think I should even consider multipart attachments - as far as I'm concerned attachments are opaque objects :)

I'll update tonight, and you can tell me how useful, or not, the idea is.  But I think it should be an improvement on what we have, and I can already imagine cutting out most of the current "get message for display" logic, and allowing each part to be viewed in-turn.

**Comment by skx** [comment]
I've spent the past hour watching The Simpsons, and hooking up this commit:

* https://github.com/skx/lumail2/commit/693a459e59fdfbf095c2c7c378376ee35f4cf363

This is the kind of thing that I had in mind, as it allows easy iteration over the available parts, and makes no obvious distinction between inline & attachments.

I would like to rework it to make use of the `CMessagePart` objects, but i was in a hurry.

With this scheme in mind I think the main GUI-driver would be simplified such that the message displayed was fetched via:

     function get_display_message( selected )
        ... get parts ..
        return the first text/plain part.  probably
        if not try to use the first text/html
        ...
        if not read the selected:path() file from disk and return literally.
     end

__That__ would remove logic from the core, allow the Lua to influence things neatly via GPG-signing, and filter-executions, etc.

That's the kind of thing I have in mind when I say "everything is an object".

Comments welcome on my approach, or implementation.  Obviously a quick hack but I think the core is solid, and it can be prototyped here in isolation.  If that works out well then the code can either be moved: lumail2 -> lumail, or a user-interface added on the former project.

Now beer.

**Comment by jugglerchris** [comment]
The multipart containers do still need to be exposed!

A message could be something like:

multipart/**mixed**  -- Several parts of the message
  * text/plain     -- dunno, some kind of header
  * multipart/**alternative**  -- Second part, which has two options depending on client capabilities
    * text/html    -- ignore, since we prefer text/plain alternative
    * text/plain   -- show this (better than the other alternative)
  * text/plain     -- show this as well (eg a signature added by mailing list)

Hiding the multipart containers and flattening the rest to [text/plain, text/html, text/plain, text/plain] has lost the information required to show the message correctly.

**Comment by skx** [comment]
I admit I'm not 100% familiar with how this stuff works, but I think the current implementation I committed doesn't flatten things.  If it does then that's a mistake, and it should be fixed.

(You'll see I only tested against two simple messages, one with text/plain + text/html, and one with a bunch of attachments.)

**Comment by jugglerchris** [comment]
I think you're missing the multipart containers because of the `GMIME_IS_PART()` test.  The containers are a GMimeMultipart instead of GMimePart (except for the top-level GMimeMessage).

**Comment by skx** [comment]
Ahah.  That was non-obvious to me.

I'll attempt to explore this shortly, unless you beat me to it with a pull-request.

**Comment by skx** [comment]
I've had a bit of a struggle with nested parts, but the good news is that a lot of the other GUI work is hooked up via lua now.

I'll leave discussion for that project in its own set of issues though:

https://github.com/lumail/lumail2/

**Comment by skx** [comment]
I'm sorry for the lack of update on this issue, but work on lumail has been pretty much abandoned.  

The good news is that there is a new project, [lumail2](https://github.com/lumail/lumail2/) which is very similar to this one:

* It is still console-based.
* It still operates solely on Maildir hierarchies.
* It is still written in C++, with all the scripting support provided by Lua.

Why change?  Lots of reasons, but primarily because I wanted everything to be an object: Messages, Maildirs, etc.  The rewrite is more functional than the older-codebase, smaller, simpler, and much more flexible.

If you're still interested in Lumail please do take a look:

* https://github.com/lumail/lumail2/
